tools/shared.py

```diff
@@ -2555,19 +2555,29 @@ def add_to_path(dirname):
       logger.debug('closure compiler: ' + ' '.join(cmd))
 
       proc = run_process(cmd, stderr=PIPE, check=False, env=env)
-      if proc.returncode != 0:
-        sys.stderr.write(proc.stderr)
-        hint = ''
-        if not pretty:
-          hint = ' the error message may be clearer with -g1'
-        exit_with_error('closure compiler failed (rc: %d.%s)', proc.returncode, hint)
 
       # XXX Closure bug: if Closure is invoked with --create_source_map, Closure should create a
       # outfile.map source map file (https://github.com/google/closure-compiler/wiki/Source-Maps)
       # But it looks like it creates such files on Linux(?) even without setting that command line
       # flag (and currently we don't), so delete the produced source map file to not leak files in
       # temp directory.
       try_delete(outfile + '.map')
+
+      if len(proc.stderr.strip()) > 0:
+        logger.debug(open(filename, 'r').read())
+
+        if proc.returncode != 0:
+          logger.error('Closure compiler run failed:\n')
+          logger.error(proc.stderr)
+          exit_with_error('closure compiler failed (rc: %d.%s)', proc.returncode, '' if pretty else ' the error message may be clearer with -g1 and EMCC_DEBUG=1 set')
```

Comment:
> How about moving this bloc up outside the `len(proc.stderr.strip()) > 0` check?

Yeah, looks good.

> How common are warning from closure for normal users?

Very common, they happen all the time.

> Does any of stdlib cause any?

Currently yes, we will need to fix so that there won't be any.

> Under what circumstances does closure choose to generate warnings but not exit with a non-zero error code?

It looks like if there are only warnings, Closure will return with a zero return code.